src/Domain/WeightedElementDistribution.cpp

```diff
@@ -0,0 +1,240 @@
+            // this proc and don't add the current one
+            add_more_elements_to_proc = false;
+          } else {
+            // otherwise, the current proc cost is farther from the target then
+            // if we were to add one more element, so we add the current element
+            // to the current proc
+            cost_spent_on_proc += element_cost;
+            cost_remaining -= element_cost;
+            num_elements_distributed_to_proc++;
+            total_elements_distributed_to_proc++;
+            current_element_of_current_block++;
+          }
+        }
+      }
+
+      // add a proc and its element allowance for the current block
+      block_element_distribution_.at(current_block)
+          .emplace_back(std::make_pair(global_proc_number,
+                                       num_elements_distributed_to_proc));
+      if (current_element_of_current_block >= num_elements_current_block) {
+        // if we're done assigning elements from the current block, move on to
+        // the elements in the next block
+        ++current_block;
+        current_element_of_current_block = 0;
+      }
+    }
+  }
+}
+
+template <size_t Dim>
+std::vector<std::vector<double>>
+WeightedBlockZCurveProcDistribution<Dim>::get_cost_by_element_by_block(
+    const std::vector<Block<Dim>>& blocks,
+    const std::vector<std::array<size_t, Dim>>& initial_refinement_levels,
+    const std::vector<std::array<size_t, Dim>>& initial_extents,
+    const Spectral::Quadrature quadrature) {
+  // elemental costs correspond to elements in Z-curve order
+  std::vector<std::vector<double>> cost_by_element_by_block(blocks.size());
+
+  for (size_t block_number = 0; block_number < blocks.size(); block_number++) {
+    const auto& block = blocks[block_number];
+    const auto initial_ref_levs = initial_refinement_levels[block.id()];
+    const std::vector<ElementId<Dim>> element_ids =
+        initial_element_ids_in_z_curve_order(block.id(), initial_ref_levs);
+    const size_t grid_points_per_element = alg::accumulate(
+        initial_extents[block.id()], 1_st, std::multiplies<size_t>());
+
+    cost_by_element_by_block[block_number].reserve(element_ids.size());
+
+    // compute the minimum grid spacing (in Frame::Grid) and cost of each
+    // element
+    for (const auto& element_id : element_ids) {
+      Mesh<Dim> mesh = ::domain::Initialization::create_initial_mesh(
+          initial_extents, element_id, quadrature);
+      Element<Dim> element = ::domain::Initialization::create_initial_element(
+          element_id, block, initial_refinement_levels);
+      ElementMap<Dim, Frame::Grid> element_map{
+          element_id, block.is_time_dependent()
+                          ? block.moving_mesh_logical_to_grid_map().get_clone()
+                          : block.stationary_map().get_to_grid_frame()};
+
+      tnsr::I<DataVector, Dim, Frame::ElementLogical> logical_coords{};
+      domain::Tags::LogicalCoordinates<Dim>::function(
```

Comment:
Can you not just call the free function here rather than the function in the tag?

Reply:
Yes, the reason I didn't call the function was just because I felt that going through the tag gave it a bit more clarity what compute item I was computing so that someone reading could quickly associate the logic to the tag. I don't have a strong opinion on this, though

Reply:
The free function is called `logical_coordinates` and has its own docs (the tag docs actually aren't super helpful). I don't think using the tag makes it clearer in this case.